File: leetcode/276.paint-fence.cpp
// f[i] = (k - 1) * f[i - 1] + (k - 1) * f[i - 2]

class Solution {
public:
    int numWays(int n, int k) {

        if (n == 0 || k == 0)
        {
            return 0;
        }

        if (n == 1)
        {
            return k;
        }

        vector<int> table(n + 1, 0);
        
        table[1] = k;
        table[2] = k * k;
        
        for (auto i = 3; i <= n; ++i)
        {
            table[i] = (k - 1) * table[i - 1] + (k - 1) * table[i - 2];
        }
        
        return table[n];
        
    }
};


class Solution2 {
public:
    int numWays(int n, int k) {
        
        if (n == 0 || k == 0)
        {
            return 0;
        }

        if (n == 1)
        {
            return k;
        }
        
        int same = k;
        int diff = k * (k - 1);
        
        for (auto i = 3; i <= n; ++i)
        {
            int tmp = diff;
            diff = (same + diff) * (k - 1);
            same = tmp;
        }
        
        return same + diff;
    }
};

File: leetcode/276.paint-fence.py
# f[i] = (k - 1) * f[i - 1] + (k - 1) * f[i - 2]
class Solution(object):
    def numWays(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: int
        """
        
        if n == 0 or k == 0:
            return 0
        
        if n == 1:
            return k
        
        f = [0 for _ in range(n + 1)]
        f[1] = k
        f[2] = k * k
        
        for i in range(3, n + 1):
            f[i] = (k - 1) * f[i - 1] + (k - 1) * f[i - 2]
            
        return f[n]

class Solution2(object):
    def numWays(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: int
        """
        
        if n == 0 or k == 0:
            return 0
        
        if n == 1:
            return k
        
        same = k
        diff = k * (k - 1)
        
        for i in range(3, n + 1):
            same, diff = diff, (same + diff) * (k - 1)
            
        return same + diff